src/main/java/com/synacy/gradprogram/spock/exercise/OrderingService.java

```diff
@@ -67,7 +72,25 @@ public Order createAnOrder(Cart cart, String recipientName, String recipientAddr
   }
 
   public void cancelOrder(CancelOrderRequest request) {
-    // TODO: Implement me. Cancels PENDING and FOR_DELIVERY orders and create a refund request saving it to the database.
-    //  Else throws an UnableToCancelException
+    String invalidOrderStatusMessage = "Cannot cancel order. Only PENDING or FOR_DELIVERY orders can be cancelled.";
+    String orderIdNotFoundMessage = "Cannot cancel order. Order ID not found.";
+
+    Optional<Order> optionalOrder = orderRepository.fetchOrderById(request.getOrderId());
+    if (optionalOrder.isPresent()) {
```

Comment:
You can simplify/improve the code's readability by utilizing the Java Streams or maybe having something like this:
```
public void cancelOrder(CancelOrderRequest request) {
  Optional<Order> optionalOrder = orderRepository.fetchOrderById(request.getOrderId());

  if (optionalOrder.isEmpty() || isOrderEligibleForCancellation(optionalOrder.get())) {
    throw new UnableToCancelException("Unable to cancel order.");
  } else {
    Order order = optionalOrder.get();
    order.setStatus(OrderStatus.CANCELLED);
    orderRepository.saveOrder(order);

    refundService.createAndSaveRefundRequest(request);
  }
}

private boolean isOrderEligibleForCancellation(Order order) {
  return order.getStatus() == OrderStatus.DELIVERED || order.getStatus() == OrderStatus.CANCELLED;
}
```